### ISSUE TITLE
Cannot read property 'type' of undefined

### ISSUE THREAD

**Issue by alexanderwe** [mariotacke/redux-electron-ipc#12]
At first thank you for this great package :) 
I have a small question and I assume I am using something wrong but can not figure out what it is. I try to to send a message from **renderer** to **main** via the following method:

```
import createIpc, { send } from 'redux-electron-ipc';

const TYPE = "myReducerTYPE"

export const updateIpc = createIpc({
  myChannel: (event, data) => {
    return {
      type: TYPE,
      data: data,
    };
  },
});


export const myFunc = () => {
  send('myChannel', {}); // I only need the channel, no additional objects or something is send
};
```

When I try to do this I am getting the following error in the console and the event is not send to **main**

```
renderer.js:35543 Uncaught TypeError: Cannot read property 'type' of undefined
    at Object.dispatch (renderer.js:35543)
    at dispatch (<anonymous>:2:1620)
    at renderer.js:33112
    at Button._this.handleClick (renderer.js:50630)
    at HTMLUnknownElement.callCallback (renderer.js:15797)
    at Object.invokeGuardedCallbackDev (renderer.js:15836)
    at Object.invokeGuardedCallback (renderer.js:15693)
    at Object.invokeGuardedCallbackAndCatchFirstError (renderer.js:15707)
    at executeDispatch (renderer.js:15930)
    at Object.executeDispatchesInOrder (renderer.js:15952)
```
And the error seems to occur on those lines:
```
return function (next) {
	      return function (action) {
	        if (action.type.startsWith('@@IPC')) {
	          _electron.ipcRenderer.send.apply(_electron.ipcRenderer, [action.channel].concat(_toConsumableArray(action.args || [])));
	        }

	        return next(action);
	      };
	    };
```
Especially it seems to be `action.type.startsWith('@@IPC')`. 
If I am sending my event with `ipcRenderer.send(...` the error still occurs but the event gets send.

Do you have a suggestion what could be wrong ? 
(Maybe because one of the objects in the createIPC has the same name as the cannel I am sending to ?)
Thanks in advance

**Comment by mariotacke** [comment]
Hi Alexander - 

Are you `dispatch`-ing the `send` method? It seems, right now your function `myFunc` executes the `send` method from my library. It has to be `dispatch`-ed though for the middleware to intercept it.

From the [docs](https://github.com/mariotacke/redux-electron-ipc#window):
```
const store = createStore(..., applyMiddleware(ipc));

// send a message with arguments through the `send` utility function
store.dispatch(send('ping', 'redux', 'electron', 'ipc'));
```

The channels for sending and receiving can be the same, there is no clash there (as one sends it to ipcMain and the other to ipcRenderer). I really think the problem here is calling the send method by itself. Try `dispatch`-ing it from redux.

**Comment by alexanderwe** [comment]
Hi Mario, 

thank you for your fast reply. I missed the fact from the docs to use `send` with the store, sorry for that. But unfortunately it does not resolve the problem completely. Now my message gets send but I still have the error in my console. The whole execution "flow" is like that.

**Action (/actions/update)**
```
import { ipcRenderer } from 'electron';
import createIpc, { send } from 'redux-electron-ipc';
import { store } from '../../entry';

export const CHECK_FOR_UPDATE = 'CHECK_FOR_UPDATE';

//Incoming IPC messages
export const updateIpc = createIpc({
  checkForUpdate: (event, data) => {
    return {
      type: CHECK_FOR_UPDATE,
      data: {
        checkingForUpdate: true,
      },
    };
  },
});

//Outgoing (UPDATED)
export const checkForUpdate = () => {
  store.dispatch(send('checkForUpdate', {}));
};
```

**Reducer**
```
import { CHECK_FOR_UPDATE, UDPATE_INFO } from '../actions/update/';

const initialState = {
  checkingForUpdate: false,
  error: null,
  updateAvailable: false,
  msg: null,
};

export function update(state = initialState, action) {
  switch (action.type) {
    case CHECK_FOR_UPDATE:
      return Object.assign({}, state, {
        checkingForUpdate: action.data.checkingForUpdate,
      });
     ....
    default:
      return state;
  }
}
```

**Init Store (entry, import the store in the actions)**
```
import { createStore, applyMiddleware } from 'redux';
import { updateIpc } from './actions/update';
import app from './reducers';

export const store = createStore(
  app,
  composeEnhancers(applyMiddleware(updateIpc))),
);
```


**In a React component:**
```
import * as React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { checkForUpdate  } from '../actions/update';

const mapStateToProps = state => ({
  update: state.update,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ checkForUpdate }, dispatch);

@connect(mapStateToProps, mapDispatchToProps)
class Settings extends React.Component {
  render() {
    return (
      <div>
        <Button
          type="primary"
          loading={this.props.update.checkingForUpdate}
          onClick={this.props.checkForUpdate}
        >
          Check for updates
        </Button>
      </div>
    );
  }
}

```

I don't know exactly why your method is receiving an undefined action object, since I am always omitting the type in any action I wrote. 
If you want to have a look at the exact source code, it is located here: https://github.com/alexanderwe/checksum-validator/tree/feature/antdUI/app

Thanks for your time and patience :)

**Comment by mariotacke** [comment]
Looks like you are not dispatching properly in your application. I looked at `/app/renderer/src/actions/update/index.js`:

```
export const checkForUpdate = () => {
  store.dispatch(send('checkForUpdate', {}));
};
```

should be:

```
export const checkForUpdate = () => {
  return send('checkForUpdate');
};
```

Generally, you don't want to re-import your store object in your action creators. However, the real problem here is that your `checkForUpdate` function does not return an action, it tries to dispatch `send` directly. This conflicts with your `Settings.component.tsx` `mapDispatchToProps`:

```
const mapDispatchToProps = dispatch =>
  bindActionCreators({ checkForUpdate, updateApp }, dispatch);
```

What happens here is that your component bindings call `checkForUpdate` on button click, but the action creator invokes dispatch itself. The return type is undefined, yet we try to dispatch it again through the redux bindings. Essentially: Button click => dispatch => store.dispatch => send.

To make it even more verbose, I would remove `bindActionCreators` altogether and replace your `mapDispatchToProps` with the following (The two versions produce the same result but show how dispatch is applied):

```
const mapDispatchToProps = (dispatch) => {
  return {
    checkForUpdate: () => dispatch(send('checkForUpdate')),
  };
}
```

Let me know if that works for you.

**Comment by mariotacke** [comment]
As a side note:
- Try creating the ipc middleware in a separate file (`ipc.js`?) instead of within your action creators. The reason here is that `createIpc` could bind multiple different action creators from different files together.
- Use `send` directly and don't wrap it into a separate function. I think it makes reasoning about it much simpler. Libraries such as `react-router-redux` have dedicated `push` methods which are dispatched directly as well. (The way `push` works was actually the inspiration for this middleware)
- `send` accepts zero or more arguments. You do not have to pass `{}` when you `send`. The first parameter (channel) is the only required parameter.
- Create a dedicated `store.js` which exports a `configureStore` function. This function will handle any middleware setup and abstract it away from your `entry.tsx`. `entry.tsx` would only call configureStore which returns the store.

**Comment by alexanderwe** [comment]
Thanks a lot for your help!  Now everything is working fine. You were right that the problem was with the `bindActionCreators` function. I changed it to you suggestion and everything is smooth. 

Additionally I adopted your suggestions and created a dedicated `ipc.js` and `store.js` but I still wrapped `send()` into separate functions, this makes it easier for me to get an overview about everything. Thanks for this suggestions, will try to keep this pattern.

I close the issue, since it is resolved :)
